Mathlib/AlgebraicGeometry/PrimeSpectrum/TensorProduct.lean

```diff
@@ -0,0 +1,78 @@
+/-
+Copyright (c) 2024 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.RingTheory.SurjectiveOnStalks
+import Mathlib.AlgebraicGeometry.PrimeSpectrum.Basic
+
+/-!
+
+# Lemmas regarding the prime spectrum of tensor products
+
+## Main result
+- `PrimeSpectrum.embedding_tensorProductTo_of_surjectiveOnStalks`:
+  If `R →+* T` is surjective on stalks (see Mathlib/RingTheory/SurjectiveOnStalks.lean),
+  then `Spec(S ⊗[R] T) → Spec S × Spec T` is a topological embedding
+  (where `Spec S × Spec T` is the cartesian product with the product topology).
+-/
+
+variable (R S T) [CommRing R] [CommRing S] [Algebra R S]
+variable [CommRing T] [Algebra R T] [Algebra S T] [IsScalarTower R S T]
+
+open TensorProduct
+
+/-- The canonical map from `Spec(S ⊗[R] T)` to the cartesian product `Spec S × Spec T`. -/
+noncomputable
+def PrimeSpectrum.tensorProductTo (x : PrimeSpectrum (S ⊗[R] T)) :
+    PrimeSpectrum S × PrimeSpectrum T :=
+  ⟨comap (algebraMap _ _) x, comap Algebra.TensorProduct.includeRight.toRingHom x⟩
+
+lemma PrimeSpectrum.continuous_tensorProductTo : Continuous (tensorProductTo R S T) :=
+  (comap _).2.prod_mk (comap _).2
+
+variable (hRT : (algebraMap R T).SurjectiveOnStalks)
+
+lemma PrimeSpectrum.embedding_tensorProductTo_of_surjectiveOnStalks_aux
+    (p₁ p₂ : PrimeSpectrum (S ⊗[R] T))
+    (h : tensorProductTo R S T p₁ = tensorProductTo R S T p₂) :
+    p₁ ≤ p₂ := by
+  let g : T →+* S ⊗[R] T := Algebra.TensorProduct.includeRight.toRingHom
+  intros x hxp₁
+  by_contra hxp₂
+  obtain ⟨t, r, a, ht, e⟩ := hRT.exists_mul_eq_tmul x
+    (p₂.asIdeal.comap g) inferInstance
+  have h₁ : a ⊗ₜ[R] t ∈ p₁.asIdeal := e ▸ p₁.asIdeal.mul_mem_left (1 ⊗ₜ[R] (r • t)) hxp₁
+  have h₂ : a ⊗ₜ[R] t ∉ p₂.asIdeal := e ▸ p₂.asIdeal.primeCompl.mul_mem ht hxp₂
+  rw [← mul_one a, ← one_mul t, ← Algebra.TensorProduct.tmul_mul_tmul] at h₁ h₂
+  have h₃ : t ∉ p₂.asIdeal.comap g := fun h ↦ h₂ (Ideal.mul_mem_left _ _ h)
+  have h₄ : a ∉ p₂.asIdeal.comap (algebraMap S (S ⊗[R] T)) :=
+    fun h ↦ h₂ (Ideal.mul_mem_right _ _ h)
+  replace h₃ : t ∉ p₁.asIdeal.comap g := by
+    rwa [show p₁.asIdeal.comap g = p₂.asIdeal.comap g from congr($h.2.1)]
```

Comment:
What does the `$` do? (personal curiosity).

Reply:
It is the syntax of `congr`. If `e` is a term of type `a = b`, `congr(f $e)` is a term of type `f a = f b`. Here it is `congr(($h).2.1)` Apparently lean is smart enough to figure out the dot notations.